### PR TITLE
Implement secret-rotate hook

### PR DESCRIPTION
**Pull request by wallyworld**
This PR wires up the previously added secret rotation worker to trigger the execution of a new `secret-rotate` hook whenever a secret should be rotated. Whenever a secret is successfully rotated, the unit agent  will, during hook commit, make a new `SecretsRotated` API call to record the last rotation time. This allows the secrets rotation worker to re-trigger a rotate event if the unit agent crashes, whether or not leadership changes before the agent restarts (only the unit leader runs the rotate hook). Secret rotation takes precedence over running other hooks like config-changed etc.

So the main components are:
- SecretsRotated api 
- new secrets hook type
- new secrets hook resolver

There's a couple of other drive by fixes:
- wire up feature flags to k8s unit agents
- adjust the valid secret URL syntax to allow secret paths with "-"
- tweak GetSecretValue API params to accept either an ID or URL

## QA steps

boostrap and deploy a charm with a simple `secret-rotated` hook, eg
```
#!/bin/bash
juju-log "secret-rotate $JUJU_SECRET_URL"
exit 0
```
create a secret without any rotate interval
`juju exec --unit ubuntu/0 "secret-create password hello=world"`

`juju show-status-log ubuntu/0` will show no secret hook runs

update the rotate interval
`juju exec --unit ubuntu/0 "secret-update password --rotate=1m"`

now observe the `secret-rotate` hook running every minute
```
$ juju show-status-log ubuntu/0
Time                        Type       Status       Message
...
17 Sep 2021 07:47:04+10:00  juju-unit  idle         
17 Sep 2021 07:48:05+10:00  juju-unit  executing    running secret-rotate hook for secret://app/ubuntu/password
17 Sep 2021 07:48:05+10:00  juju-unit  idle         
17 Sep 2021 07:49:06+10:00  juju-unit  executing    running secret-rotate hook for secret://app/ubuntu/password
17 Sep 2021 07:49:06+10:00  juju-unit  idle 
```

juju debug log shows the hook was run with the correct secret url in the context
```
...
unit-ubuntu-0: 07:48:05 INFO unit.ubuntu/0.juju-log secret-rotate secret://app/ubuntu/password
...
```

disable rotate interval and secret-rotate hook no longer runs
`juju exec --unit ubuntu/0 "secret-update password --rotate=0"`

### CODE REVIEW

**Comment by wallyworld** [comment]
[secret-rotate hook implementation](https://trello.com/c/Ko8Lidnn/1755-secret-rotate-hook-implementation)

File: worker/uniter/resolver_test.go [hunk 1]
```diff
@@ -9,6 +9,7 @@ import (
 	"github.com/juju/charm/v8"
 	"github.com/juju/charm/v8/hooks"
 	"github.com/juju/errors"
+	"github.com/juju/juju/worker/uniter/secrets"
```
Comment (hpidcock) [inline worker/uniter/resolver_test.go:12]: wrong block

File: worker/uniter/runner/context/env_test.go [hunk 2]
```diff
@@ -8,6 +8,7 @@ import (
 	"runtime"
 	"sort"
 
+	"github.com/juju/juju/core/secrets"
```
Comment (hpidcock) [inline worker/uniter/runner/context/env_test.go:11]: wrong block

**Review by hpidcock — APPROVED** [review]
LGTM

File: api/secretsmanager/client.go [hunk 3]
```diff
@@ -71,7 +74,11 @@ func (c *Client) Create(cfg *secrets.SecretConfig, secretType secrets.SecretType
 }
 
 // Update updates an existing secret value and/or config like rotate interval.
-func (c *Client) Update(URL *secrets.URL, cfg *secrets.SecretConfig, value secrets.SecretValue) (string, error) {
+func (c *Client) Update(url string, cfg *secrets.SecretConfig, value secrets.SecretValue) (string, error) {
+	URL, err := secrets.ParseURL(url)
+	if err != nil {
+		return "", errors.Trace(err)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply api/secretsmanager/client.go:80]: the underlying error from ParseURL() is already a NotValid error with a description

File: worker/uniter/hook/hook.go [hunk 4]
```diff
@@ -62,46 +59,53 @@ func (hi Info) Validate() error {
...
-	// TODO(fwereade): define these in charm/hooks...
-	case LeaderElected, LeaderDeposed, LeaderSettingsChanged:
+	case hooks.LeaderElected, hooks.LeaderDeposed, hooks.LeaderSettingsChanged:
+		return nil
+	case hooks.SecretRotate:
+		if hi.SecretURL == "" {
+			return errors.Errorf("%q hook requires a secret URL", hi.Kind)
+		}
+		if _, err := secrets.ParseURL(hi.SecretURL); err != nil {
+			return errors.Errorf("invalid secret URL %q", hi.SecretURL)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply worker/uniter/hook/hook.go:104]: am following existing convention here.  ie there's a whole bunch of error messages like
`return errors.Errorf("invalid storage ID %q", hi.StorageId)`

Maybe consider changing holistically in another PR

File: worker/uniter/op_callbacks.go [hunk 5]
```diff
@@ -159,3 +160,8 @@ func (opc *operationCallbacks) RemoteInit(runningStatus remotestate.ContainerRun
 	}
 	return opc.u.remoteInitFunc(runningStatus, abort)
 }
+
+// SetSecretRotated is part of the operation.Callbacks interface.
+func (opc *operationCallbacks) SetSecretRotated(url string, when time.Time) error {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply worker/uniter/op_callbacks.go:165]: Yep, there's a TODO somewhere to that effect

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**